hack/build-and-push.sh

```diff
@@ -133,95 +180,324 @@ if [ "$SKIP_BUILD" == "" ]; then
+    cp "$task_file" "$prepared_task_file"
+    local -r task_file_sha=$(git log -n 1 --pretty=format:%H -- "$task_file")
+    local -r task_bundle=$(generate_tagged_task_bundle "$task_name" "$task_version")
+    echo "$prepared_task_file $task_file_sha $task_bundle"
+}
+
+# Generate build data for a normal task. The data is output to stdout as
+# space-separated fields in a single line.
+# Arguments: task_dir, task_name, task_version
+generate_kustomized_task_build_data() {
+    local -r task_dir=$1 task_name=$2 task_version=$3
+    local -r task_file="$task_dir/$task_name.yaml"
+    local -r prepared_task_file="${WORKDIR}/${task_name}-${task_version}.yaml"
+    oc kustomize "$task_dir" >"$prepared_task_file"
+    local -r task_file_sha=$(sha256sum "$prepared_task_file" | awk '{print $1}')
+    local -r task_bundle=$(generate_tagged_task_bundle "$task_name" "$task_version")
+    echo "$prepared_task_file $task_file_sha $task_bundle"
+}
+
+declare -r ENABLE_CACHE=${ENABLE_CACHE:-""}
+CACHE_FILE=
+
+if [ -n "$ENABLE_CACHE" ]; then
+    CACHE_FILE="/tmp/build-definitions-build-and-push.cache"
+    declare -r CACHE_FILE
+fi
+
+cache_get() {
+    local -r key=$1
+    local value=
+    if [ -n "$ENABLE_CACHE" ]; then
+        # No match should not fail the script. It means no cached data.
+        read -r _ value < <(grep "^${key}" "$CACHE_FILE" 2>/dev/null) || :
+    fi
+    echo "$value"
+}
+
+cache_set() {
+    local -r key=$1 value=$2
+    if [ -n "$ENABLE_CACHE" ]; then
+        echo "${key} ${value}" >>"$CACHE_FILE"
+    fi
+}
+
+# Fetch image digest.
+# Arguments: image
+# The digest is output to stdout.
+fetch_image_digest() {
+    local -r image=$1
+    local digest=
+    digest=$(cache_get "$image")
+    if [ -z "$digest" ]; then
+        digest=$(skopeo inspect --no-tags --format='{{.Digest}}' "docker://${image}" 2>/dev/null)
+        if [ -n "$digest" ]; then
+            cache_set "$image" "$digest"
         fi
-        if [ -f "${task_dir}/TROUBLESHOOTING.md" ]; then
-            ANNOTATIONS+=("dev.tekton.docs.troubleshooting=${VCS_URL}/tree/${VCS_REF}/${task_dir}TROUBLESHOOTING.md")
+    fi
+    echo "$digest"
+}
+
+# Attach migration file to given task bundle.
+# Arguments: task_dir, concrete_task_version, task_bundle
+attach_migration_file() {
+    local -r task_dir=$1
+    local -r concrete_task_version=$2
+    local -r task_bundle=$3
+
+    local -r task_file="${task_dir}/${task_name}.yaml"
+    local -r migration_file="${task_dir}/migrations/${concrete_task_version}.sh"
+
+    if [ ! -e "$migration_file" ]; then
+        return 0
+    fi
+
+    # Check if task bundle has an attached migration file.
+    local filename
+    local found=
+    local artifact_refs
+
+    # List attached artifacts, that have specific artifact type and annotation.
+    # Then, find out the migration artifact.
+    artifact_refs=$(
+        retry oras discover "$task_bundle" --artifact-type "$ARTIFACT_TYPE_TEXT_XSHELLSCRIPT" --format json | \
```

Comment:
The `--format` flag is new in https://github.com/oras-project/oras/releases/tag/v1.2.0

It would be good to document this or even add an automated check that the version is new enough. And we need to make sure that the appstudio-utils image has a new enough oras installed.